src/main/scala/mesosphere/marathon/storage/migration/MigrationTo18.scala

```
@@ -0,0 +1,175 @@
+      (__ \ "unreachableStrategy").readNullable[raml.UnreachableStrategy] ~
+      (__ \ "reservation").readNullable[Reservation]
+    ) { (instanceId, agentInfo, tasksMap, runSpecVersion, state, ramlUnreachableStrategy, reservation) =>
+        val unreachableStrategy = ramlUnreachableStrategy.
+          map(Raml.fromRaml(_)).getOrElse(UnreachableStrategy.default())
+        new Instance(instanceId, Some(agentInfo), state, tasksMap, runSpecVersion, unreachableStrategy, reservation)
+      }
+  }
+
+  implicit val instanceResolver: IdResolver[Instance.Id, JsValue, String, ZkId] =
+    new IdResolver[Instance.Id, JsValue, String, ZkId] {
+      override def toStorageId(id: Id, version: Option[OffsetDateTime]): ZkId =
+        ZkId(category, id.idString, version)
+
+      override val category: String = "instance"
+
+      override def fromStorageId(key: ZkId): Id = Instance.Id.fromIdString(key.id)
+
+      override val hasVersions: Boolean = false
+
+      override def version(v: JsValue): OffsetDateTime = OffsetDateTime.MIN
+    }
+
+  implicit val instanceJsonUnmarshaller: Unmarshaller[ZkSerialized, JsValue] =
+    Unmarshaller.strict {
+      case ZkSerialized(byteString) =>
+        Json.parse(byteString.utf8String)
+    }
+
+  /**
+    * This function traverses all instances in ZK and sets the instance goal field.
+    */
+  def migrateInstanceConditions(instanceRepository: InstanceRepository, persistenceStore: PersistenceStore[_, _, _])(implicit mat: Materializer): Future[Done] = {
+
+    logger.info("Starting instance condition migration")
+
+    maybeStore(persistenceStore).map { store =>
+      instanceRepository
+        .ids()
+        .mapAsync(1) { instanceId =>
+          store.get[Instance.Id, JsValue](instanceId)
+        }
+        .via(migrationFlow)
+        .mapAsync(1) { updatedInstance =>
+          instanceRepository.store(updatedInstance)
+        }
+        .runWith(Sink.ignore)
+    } getOrElse {
+      Future.successful(Done)
```

Shouldn't we rather throw an exception if no instance persistence store was available?

---

It's not the exception handling, this branch will be executed in case the migration is done on InMemoryStore